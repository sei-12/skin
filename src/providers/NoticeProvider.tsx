import { Alert, Snackbar, SnackbarCloseReason } from "@mui/material";
import React, { createContext, useCallback, useContext, useState } from "react";

export type Severity = "info" | "warning" | "error" | "success";

export interface Notice {
    serverity: Severity;
    message: string;

    /**
     * default: 6000
     */
    timeout_ms?: number;
}

type NoticeContext = {
    addNotice: (notice: Notice) => void;
};

const noticeContext = createContext<NoticeContext>({
    addNotice: () => {},
});

export function useNotice() {
    const n = useContext(noticeContext);
    return n;
}

export function NoticeProvider({ children }: { children: React.ReactNode }) {
    const [open, setOpen] = React.useState(false);
    const [message, setMessage] = useState("");
    const [severity, setSeverity] = useState<Severity>("success");
    const [autoHideDuration, setAutoHideDuration] = useState(6000);

    const handleClose = (
        _?: React.SyntheticEvent | Event,
        reason?: SnackbarCloseReason
    ) => {
        if (reason === "clickaway") {
            return;
        }

        setOpen(false);
    };

    const addNotice = useCallback((notice: Notice) => {
        setMessage(notice.message);
        setOpen(true);
        setSeverity(notice.serverity);
        setAutoHideDuration(notice.timeout_ms ?? 6000);
    }, []);

    const [notice] = useState<NoticeContext>({
        addNotice,
    });

    return (
        <noticeContext.Provider value={notice}>
            <Snackbar
                open={open}
                autoHideDuration={autoHideDuration}
                onClose={handleClose}
            >
                <Alert
                    onClose={handleClose}
                    severity={severity}
                    variant="filled"
                    sx={{ width: "100%" }}
                >
                    {message}
                </Alert>
            </Snackbar>
            {children}
        </noticeContext.Provider>
    );
}
