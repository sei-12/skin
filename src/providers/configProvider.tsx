/* eslint-disable react-refresh/only-export-components */

import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import { createContext, useContext, useEffect, useState } from "react";
import type { Config } from "../../src-tauri/bindings/export/Config";

export function useConfig() {
    const c = useContext(ConfigContext);
    return c;
}

export function ConfigProvider({ children }: { children: React.ReactNode }) {
    const [config, setConfig] = useState<Config>(DEFAULT_CONFIG);

    useEffect(() => {
        invoke<Config>("get_config").then((con) => {
            setConfig(con);
        });

        listen("change-config-file", () => {
            invoke<Config>("get_config").then((con) => {
                setConfig(con);
            });
        });
    }, []);

    return (
        <ConfigContext.Provider value={config}>
            {children}
        </ConfigContext.Provider>
    );
}

// テストで使うためにexport
export const DEFAULT_CONFIG: Config = {
    colorTheme: {
        bg: "#000001",
        bookmarkItem: {
            bg: "#000002",
            desc: "#000003",
            title: "#000004",
            focusBg: "#000005",
            tag: "#000006",
        },
        suggestionWindow: {
            bg: "#000007",
            borderColor: "#000008",
            focusBg: "#000009",
            match: "#000010",
            unmatch: "#000011",
        },
        predicateInputBox: {
            bg: "#000012",
            caretColor: "#000013",
            textColor: "#000014",
            placeholder: "#000015",
        },
        tagItem: {
            bg: "#000013",
            borderColor: "#000014",
            exists: "#000015",
            notExists: "#000016",
        },
        createNewBookmark: {
            caretColor: "white",
            placeholder: "#000015",
            textColor: "white",
        }
    },
    keybinds: {
        focusDownBookmarkList: ["ctrl+n", "ArrowDown"],
        focusUpBookmarkList: ["ctrl+p", "ArrowUp"],
        closeWindow: "Escape",
        focusDownSuggestionWindow: ["ctrl+n", "ArrowDown"],
        focusUpSuggestionWindow: ["ctrl+p", "ArrowUp"],
        addFocusedSuggestionItem: "Enter",
        popInputedTag: "Backspace",
        closeSuggestionWindow: "Escape",
        navigateCreateNewBookmark: "ctrl+a",
        openUrl: "Enter",
        cancelCreateNewBookmark: "Escape",
        doneCreateNewBookmark: "ctrl+Enter",
        takeInputTag: "Space",
    },
};

const ConfigContext = createContext<Config>(DEFAULT_CONFIG);
