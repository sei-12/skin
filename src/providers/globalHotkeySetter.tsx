/**
 * プロバイダーじゃないけどどこにおいたらいいか分からないし、新しいフォルダを作るのは違うと思った。
 * 一番性質的に似てる気がするからプロバイダーにおく
 * もしフォルダ分けするならエフェクターとか?
 */

import { useEffect } from "react";
import { useConfig } from "./configProvider";
import { register, unregisterAll } from "@tauri-apps/plugin-global-shortcut";
import { WindowVisibleController } from "../services/windowVisibleController";

export function GlobalHotkeySetter() {
    const config = useConfig();

    const reset = async () => {
        await unregisterAll().catch((e) => {
            console.error(e);
        });

        await register(config.keybinds.global.toggleWindowVisible, (e) => {
            if (e.state === "Released") {
                return;
            }
            WindowVisibleController.toggle();
        }).catch((e) => {
            console.warn("理由が分からない", e);
        });
    };

    useEffect(() => {
        reset();
    }, [config]);

    return <> </>;
}
