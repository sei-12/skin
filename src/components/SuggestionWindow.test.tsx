import {
    SuggestionWindow,
    SuggestionWindowItem,
    SuggestionWindowItemTextBlock,
} from "./SuggestionWindow";
import { render, renderHook, screen } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { describe, expect, test, vi } from "vitest";
import { useRef } from "react";
import { DEFAULT_CONFIG } from "../providers/configProvider";
import userEvent from "@testing-library/user-event";

describe("SuggestionWindow", () => {
    test("items", () => {
        const items: [string, boolean][][] = [
            [
                ["h", true],
                ["ello", false],
            ],
        ];
        const focusIndex = 0;
        const colorTheme = DEFAULT_CONFIG.colorTheme;
        const refs = renderHook(() => {
            return useRef<(HTMLDivElement | null)[]>([]);
        });

        const { container } = render(
            <SuggestionWindow
                items={items}
                onClickItem={vi.fn()}
                focusIndex={focusIndex}
                itemRefs={refs.result.current}
                colorTheme={colorTheme}
            ></SuggestionWindow>,
        );

        expect(container).matchSnapshot();

        const matchBlock = screen.getByText("h");
        expect(matchBlock).toBeInTheDocument();
        expect(matchBlock).toHaveStyle(
            "color: " + colorTheme.suggestionWindow.match + ";",
        );
        expect(matchBlock).toMatchSnapshot();

        const unmatchBlock = screen.getByText("ello");
        expect(unmatchBlock).toBeInTheDocument();
        expect(unmatchBlock).toHaveStyle(
            "color: " + colorTheme.suggestionWindow.unmatch + ";",
        );
        expect(unmatchBlock).toMatchSnapshot();
    });

    test("test2", () => {
        const items: [string, boolean][][] = [[["helloworld", true]]];
        const focusIndex = 0;
        const colorTheme = DEFAULT_CONFIG.colorTheme;
        const refs = renderHook(() => {
            return useRef<(HTMLDivElement | null)[]>([]);
        });

        render(
            <SuggestionWindow
                items={items}
                onClickItem={vi.fn()}
                focusIndex={focusIndex}
                itemRefs={refs.result.current}
                colorTheme={colorTheme}
            ></SuggestionWindow>,
        );

        const matchBlock = screen.getByText("helloworld");
        expect(matchBlock).toBeInTheDocument();
        expect(matchBlock).toHaveStyle(
            "color: " + colorTheme.suggestionWindow.match + ";",
        );
    });

    test("test3", () => {
        const items: [string, boolean][][] = [[["helloworld", false]]];
        const focusIndex = 0;
        const colorTheme = DEFAULT_CONFIG.colorTheme;
        const refs = renderHook(() => {
            return useRef<(HTMLDivElement | null)[]>([]);
        });

        render(
            <SuggestionWindow
                items={items}
                onClickItem={vi.fn()}
                focusIndex={focusIndex}
                itemRefs={refs.result.current}
                colorTheme={colorTheme}
            ></SuggestionWindow>,
        );

        const unmatchBlock = screen.getByText("helloworld");
        expect(unmatchBlock).toBeInTheDocument();
        expect(unmatchBlock).toHaveStyle(
            "color: " + colorTheme.suggestionWindow.unmatch + ";",
        );
        expect(unmatchBlock).toMatchSnapshot();
    });

    test("test3", async () => {
        const items: [string, boolean][][] = [
            [["helloworld", true]],
            [["hello", true]],
            [["aaaaa", true]],
        ];

        const focusIndex = 0;
        const colorTheme = DEFAULT_CONFIG.colorTheme;
        const refs = renderHook(() => {
            return useRef<(HTMLDivElement | null)[]>([]);
        });

        const onClickItem = vi.fn();
        render(
            <SuggestionWindow
                items={items}
                onClickItem={onClickItem}
                focusIndex={focusIndex}
                itemRefs={refs.result.current}
                colorTheme={colorTheme}
            ></SuggestionWindow>,
        );

        const user = userEvent.setup();

        await user.click(screen.getByText("helloworld"));
        expect(onClickItem).toBeCalledTimes(1);
        expect(onClickItem).toBeCalledWith(0);

        await user.click(screen.getByText("hello"));
        expect(onClickItem).toBeCalledTimes(2);
        expect(onClickItem).toBeCalledWith(1);

        await user.click(screen.getByText("aaaaa"));
        expect(onClickItem).toBeCalledTimes(3);
        expect(onClickItem).toBeCalledWith(2);

        await user.click(screen.getByText("helloworld"));
        expect(onClickItem).toBeCalledTimes(4);
        expect(onClickItem).toBeCalledWith(0);
    });
});

describe("TextBlock", () => {
    test("renders with correct styles for matched text", () => {
        render(
            <SuggestionWindowItemTextBlock
                isMatch={true}
                text="Match"
                colorTheme={DEFAULT_CONFIG.colorTheme}
            />,
        );
        const textElement = screen.getByText("Match");
        expect(textElement).toBeInTheDocument();
    });

    test("renders with correct styles for unmatched text", () => {
        render(
            <SuggestionWindowItemTextBlock
                isMatch={false}
                text="Unmatch"
                colorTheme={DEFAULT_CONFIG.colorTheme}
            />,
        );
        const textElement = screen.getByText("Unmatch");
        expect(textElement).toBeInTheDocument();
    });
});

describe("Item", () => {
    test("renders item with highlighted blocks", () => {
        const item: [string, boolean][] = [
            ["a", true],
            ["a", false],
            ["b", true],
            ["b", false],
            ["c", true],
            ["c", false],
        ];
        render(
            <SuggestionWindowItem
                item={item}
                focus={false}
                onClick={() => {}}
                ref={null}
                colorTheme={DEFAULT_CONFIG.colorTheme}
            />,
        );

        expect(screen.getAllByText("a")[0]).toBeInTheDocument();
        expect(screen.getAllByText("a")[1]).toBeInTheDocument();
        expect(screen.getAllByText("b")[0]).toBeInTheDocument();
        expect(screen.getAllByText("b")[1]).toBeInTheDocument();
        expect(screen.getAllByText("c")[0]).toBeInTheDocument();
        expect(screen.getAllByText("c")[1]).toBeInTheDocument();
    });

    test("test click", async () => {
        const user = userEvent.setup();
        const item: [string, boolean][] = [
            ["a", true],
            ["a", false],
            ["b", true],
            ["b", false],
            ["c", true],
            ["c", false],
        ];

        const onClick = vi.fn();
        render(
            <SuggestionWindowItem
                item={item}
                focus={false}
                onClick={onClick}
                ref={null}
                colorTheme={DEFAULT_CONFIG.colorTheme}
            />,
        );

        expect(onClick).toBeCalledTimes(0);
        const a = screen.getByTestId("suggestion-item");
        await user.click(a);
        expect(onClick).toBeCalledTimes(1);
    });
});
