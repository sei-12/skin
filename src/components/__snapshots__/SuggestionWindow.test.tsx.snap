// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SuggestionWindow > items 1`] = `
<div>
  <div
    class="MuiBox-root css-1wyizo9"
    data-testid="suggestion-window"
  >
    <div
      class="MuiBox-root css-1hfa24n"
      data-testid="suggestion-item"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1bfdmob-MuiTypography-root"
      >
        h
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-w44nn4-MuiTypography-root"
      >
        ello
      </p>
    </div>
  </div>
</div>
`;

exports[`SuggestionWindow > items 2`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-1bfdmob-MuiTypography-root"
>
  h
</p>
`;

exports[`SuggestionWindow > items 3`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-w44nn4-MuiTypography-root"
>
  ello
</p>
`;
